plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDHighlightingManager.java

```diff
@@ -0,0 +1,194 @@
+import org.jkiss.dbeaver.erd.ui.part.AttributePart;
+import org.jkiss.dbeaver.erd.ui.part.EntityPart;
+import org.jkiss.dbeaver.utils.ListNode;
+
+
+public class ERDHighlightingManager {
+
+    private static final Log log = Log.getLog(ERDHighlightingManager.class);
+
+    @NotNull
+    private final Map<IFigure, PartHighlighter> highlightedParts = new HashMap<>();
+
+    private static final class HighlightingEntry {
+        public final Color color;
+
+        public HighlightingEntry(@NotNull Color color) {
+            this.color = color;
+        }
+    }
+
+    private final class PartHighlighter {
+        @NotNull
+        private final IFigure part;
+        @NotNull
+        private final Color originalColor;
+        private final boolean originalOpaque;
+        @NotNull
+        private final LinkedList<HighlightingEntry> highlightings = new LinkedList<>();
+
+        public PartHighlighter(@NotNull IFigure part) {
+            this.originalColor = part instanceof Connection ? part.getForegroundColor() : part.getBackgroundColor();
+            this.originalOpaque = part.isOpaque();
+            this.part = part;
+        }
+
+        private void refresh() {
+            try {
+                if (this.highlightings.isEmpty()) {
+                    if (part instanceof Connection) {
+                        part.setForegroundColor(originalColor);
+                    } else {
+                        part.setBackgroundColor(originalColor);
+                    }
+                    part.setOpaque(originalOpaque);
+                } else {
+                    if (part instanceof Connection) {
+                        part.setForegroundColor(highlightings.getLast().color);
+                    } else {
+                        part.setBackgroundColor(highlightings.getLast().color);
+                    }
+                    part.setOpaque(true);
+                }
+                part.repaint();
+            } catch (Throwable ex) {
+                // any of the figure setters may internally use repaint(), and any of them could use
+                // internal object model infrastructure of the Eclipse Graphics Editor, which might be partially invalidated
+                // due to the underlying mechanics, which are out of the scope of the highlighting logic
+                log.warn("Inconsistent highlighting management detected during figure props refresh.", ex);
+            }
+        }
+
+        @NotNull
+        public ERDHighlightingHandle highlight(@NotNull Color color) {
+            HighlightingEntry entry = new HighlightingEntry(color);
+            highlightings.addLast(entry);
+            this.refresh();
+            
+            return () -> {
+                // we are removing entry somewhere in the middle of the highlighting stack, so reference identity is required,
+                // which Color class does not follow, as it supports by-value equivalence
```

Comment:
This comment will make more sense if placed directly near the `HighlightingEntry` class.